### ISSUE TITLE
use unmarshalbinary

### ISSUE THREAD

**Issue by aviau** [BurntSushi/toml#225]
It looks line net/url only implements UnmarshalBinary:
 - https://golang.org/pkg/net/url/#URL.UnmarshalBinary

Should the toml package use it too?

BTW, I have opened a bug on golang: https://github.com/golang/go/issues/25705

Even if this is a bug with golang, maybe it wouldn't be bad to try UnmarshalBinary when UnmarshalText is not present.

**Comment by arp242** [comment]
What's the practical benefit of supporting this? What problems does it solve?

**Comment by sorcix** [comment]
Looks like the stdlib `encoding/json` package also doesn't support this, so it may not be worth it.

https://go.dev/play/p/8XfmdkXecHm

**Comment by arp242** [comment]
Yeah dunno, I need to look at it. At least I understand why it's being asked.

**Comment by cyphar** [comment]
`*"net/url".URL` (among other types) does not implement `UnmarshalText` because of concerns around backwards compatibility (see https://github.com/golang/go/issues/25705 and https://github.com/golang/go/issues/10275). Instead, they added `UnmarshalBinary` helpers that actually return and parse strings.

The net result is that if you embed a `*"net/url".URL` into a `struct`, `toml` will not be able to encode or decode it as a string, meaning that you need to create wrapper structures like:

```go
// tomlURL provides a wrapper around [url.URL] which can be parsed and
// serialised as a TOML string.
type tomlURL struct {
	*url.URL
}

// The Go stdlib does not provide this method because of concerns around
// compatibility, so we need to work around this. <https://go.dev/issue/25705>
func (u *tomlURL) MarshalText() ([]byte, error) {
	return u.MarshalBinary() // implements string-based unmarshalling
}

// The Go stdlib does not provide this method because of concerns around
// compatibility, so we need to work around this. <https://go.dev/issue/25705>
func (u *tomlURL) UnmarshalText(data []byte) error {
	return u.UnmarshalBinary(data) // implements string-based unmarshalling
}
```

However, `UnmarshalBinary` means very different things for types like hashes, so maybe it makes more sense to just special-case the set of stdlib types which implement `UnmarshalBinary` instead of `UnmarshalText` for string parsing.

FWIW, I opened a very similar issue against `pelletier/go-toml` just now as well (https://github.com/pelletier/go-toml/issues/1058).
